### PR TITLE
Documentation added

### PR DESCRIPTION
**Pull request by cmrd-senya**
Here are some documentation updates.
It's not finished yet, here I'll post further updates.

### CODE REVIEW

File: lib/diaspora_federation/entities/relayable.rb [hunk 1]
```diff
@@ -41,6 +58,11 @@ def self.verify_signatures(data)
         end
       end
 
+      # Adds signatures to a given hash with the keys of the author and the parent
+      # if the signatures are not in the hash yet and if the keys are available.
+      #
+      # @param [Hash] data hash given for a signing
+      # @return [Hash] reference to the input hash
```
Comment (diaspora-code-review) [inline lib/diaspora_federation/entities/relayable.rb]: Assignment Branch Condition size for verify_signatures is too high. [21.68/20]

**Comment by cmrd-senya** [comment]
I've added documentation to all the objects I had introduced. Only two methods left reported by yardoc:
```
Files:          81
Modules:        15 (    0 undocumented)
Classes:        84 (    0 undocumented)
Constants:      22 (    0 undocumented)
Methods:       211 (    2 undocumented)
 99.40% documented

Undocumented Objects:

(in file: lib/diaspora_federation/salmon/slap.rb)
DiasporaFederation::Salmon::Slap.build_xml
DiasporaFederation::Salmon::Slap#add_magic_env_from_doc
```

I'm unsure what to write.

I also noticed that some method are not reported by yardoc, though they don't have documentation defined.

**Comment by SuperTux88** [comment]
I can add the missing ones, and I think yardoc only shows public methods.

File: lib/diaspora_federation/entities/relayable_retraction.rb [hunk 2]
```diff
@@ -1,10 +1,33 @@
...
       property :target_type
+
+      # @!attribute [r] sender_id
+      #   The diaspora ID of the person who deletes a post
+      #   @see Person#diaspora_id
+      #   @return [String] diaspora ID
       property :sender_id, xml_name: :sender_handle
+
+      # @!attribute [r] target_author_signature
+      #   Contains a signature of the entity using the private key of the author of a post itself
```
Comment (SuperTux88) [inline lib/diaspora_federation/entities/relayable_retraction.rb]: I think this is the signature of the author of the comment/like/whatever?

Do you know what happens if the author of the parent-post deletes a comment? (the author if a post has the permissions to moderate the comments). who checks this signature and when is it needed?

I think we should document this things too, if we know it :)

File: lib/diaspora_federation/entities/relayable.rb [hunk 3]
```diff
@@ -1,10 +1,26 @@
...
+      # on inclusion of this module the required properties for a relayable are added to the object that includes it
       def self.included(model)
         model.class_eval do
+          # @!attribute [r] parent_guid
+          #   @see HCard#guid
+          #   @return [String] parent guid
           property :parent_guid
+
+          # @!attribute [r] parent_author_signature
+          #   Contains a signature of the entity using the private key of the author of a parent post
```
Comment (SuperTux88) [inline lib/diaspora_federation/entities/relayable.rb]: Can you add the info here, when which signature needs to be verified?

File: lib/diaspora_federation/entities/signed_retraction.rb [hunk 4]
```diff
@@ -1,9 +1,28 @@
...
       property :target_type
+
+      # @!attribute [r] sender_id
+      #   The diaspora ID of the person who deletes a post
+      #   @see Person#diaspora_id
+      #   @return [String] diaspora ID
       property :sender_id, xml_name: :sender_handle
+
+      # @!attribute [r] author_signature
+      #   Contains a signature of the entity using the private key of the author of a post
```
Comment (SuperTux88) [inline lib/diaspora_federation/entities/signed_retraction.rb]: when does a retraction need to be signed? (SignedRetraction vs Retraction)

File: lib/diaspora_federation/salmon/aes.rb [hunk 5]
```diff
@@ -17,7 +17,6 @@ def self.generate_key_and_iv
       # @param [String] data plain input
       # @param [String] key AES key
       # @param [String] iv AES initialization vector
-      # @param [Hash] key_and_iv { key: "...", iv: "..." }
```
Comment (SuperTux88) [inline lib/diaspora_federation/salmon/aes.rb]: thanks, nice catch, refactorings ... ;)

**Comment by SuperTux88** [comment]
I have added some questions, but if you don't know the answers I have maybe some time at the weekend to do some research :)

File: lib/diaspora_federation/entities/signed_retraction.rb (hunk 4 above)
Comment (cmrd-senya) [inline lib/diaspora_federation/entities/signed_retraction.rb]: The retractions are issued from [here](https://github.com/diaspora/diaspora/blob/develop/app/models/user.rb#L377-L383).

RelayableRetraction is used for relayables, SignedRetraction is used for posts, and Retraction is used for everything else, which is ATM includes only photo.

File: lib/diaspora_federation/entities/relayable_retraction.rb (hunk 2 above)
Comment (cmrd-senya) [inline lib/diaspora_federation/entities/relayable_retraction.rb]: Yes, it is the signature of the relayable, I must reword it.

It works pretty much the same way the relayable federation works.

When author of the parent post deletes the comment this signature [is not checked](https://github.com/diaspora/diaspora/blob/develop/lib/diaspora/federated/relayable_retraction.rb#L51-L53).

**Comment by cmrd-senya** [comment]
I added some answers on your questions. I added it as a different commit so it's easier to see the changes.

**Comment by cmrd-senya** [comment]
Actually, it seems to me that signature checks for Retractions of every kind doesn't make much sense. In fact every Retraction is proved only by the signature of its sender (it's different from the Relayable federation, because parent author is allowed to retract comments of any author). And the signature of the sender is checked at the Salmon level. So maybe abolishment of signatures for Retractions (done gradually) won't make any harm. @jhass?

In fact, for the Relayables it's only necessary to pass author signature and parent author signature is redundant for the same reasons - it just repeats Salmon signature check.

**Comment by cmrd-senya** [comment]
I think proposals in my last comment probably need an issue to be opened in the diaspora tracker or here.

**Comment by cmrd-senya** [comment]
For retractions abolishment of signatures can be done pretty softly, by just moving to Retraction.

**Comment by cmrd-senya** [comment]
I've opend an issue on the Diaspora tracker on the signatures problem. It's out of scope here anyway.

**Comment by cmrd-senya** [comment]
@SuperTux88, tell me if it's fine to squash the commits or I have to fix something else.

File: lib/diaspora_federation/entities/like.rb [hunk 6]
```diff
@@ -1,10 +1,30 @@
 module DiasporaFederation
   module Entities
+    # this entity represents a like to some kind of post (e.g. status message)
+    #
+    # @see Validators::LikeValidator
     class Like < Entity
+      # @!attribute [r] positive
+      #   If true set a like, if false, remove it
```
Comment (SuperTux88) [inline lib/diaspora_federation/entities/like.rb]: i thought positive=false is a dislike (which is not implemented, but possible by the protocol)? and removing a like is done by a retraction?

Comment (cmrd-senya) [inline lib/diaspora_federation/entities/like.rb]: True, my mistake. [Here](https://github.com/diaspora/diaspora/blob/develop/lib/diaspora/likeable.rb#L9-L10) is what hints to this.

**Comment by SuperTux88** [comment]
I think you can squash it now :)

**Comment by cmrd-senya** [comment]
done

**Comment by SuperTux88** [comment]
thanks :)

**Pull request merged**